### ISSUE TITLE
Zero-width space causes issues with Xcode

### ISSUE THREAD

**Issue by mattheworiordan** [codemirror/CodeMirror#4776]
One of our users [has reported issues copying & pasting code from our CodeMirror rendered code into Xcode](https://github.com/ably/ably-ios/issues/595#issuecomment-305395846). The issue is that CodeMirror uses `&#8203;` (a zero width space U+200B) for empty code lines, and Xcode complains about these with the error "Expected declaration"

![voila_capture 2017-06-01_10-42-35_am](https://cloud.githubusercontent.com/assets/43789/26674157/4c387574-46b7-11e7-8c97-7003ba7f7134.png)

I would be happy to try a PR to modify this, but I suspect there's good reason behind why `&#8203;` is used in CodeMirror for empty lines. Any feedback on this would be greatly appreciated.

Note, inspecting the source code of a code block like:

![image](https://cloud.githubusercontent.com/assets/43789/26674200/827f8762-46b7-11e7-8b26-28d105b9093b.png)

reveals the following code:

![image](https://cloud.githubusercontent.com/assets/43789/26674213/93e3a16e-46b7-11e7-8348-7854ee66d51d.png)

### PR details

* Browsers: Tried Chrome & Safari.  I assume this is not a browser issue though as it's the source code that is generated
* Version: 5.11

**Comment by marijnh** [comment]
This _is_ a browser issue -- CodeMirror tries to handle copy/cut events and put the appropriate text content on the clipboard. I.e. the HTML that is used to render the code should not be the stuff that ends up on the clipboard. This appears to be working in my Chrome (Linux) -- if I copy a bunch of highlighted text and paste it into a contenteditable element, I get plain text, not colored HTML.

I don't have XCode set up and I don't really feel like jumping through all the Apple hoops required to install it, so could you verify that when you paste into a contenteditable HTML input, you're also seeing the problem?

**Comment by shantamg** [comment]
I haven't tried a contenteditable HTML input, but if I paste into VIM, everything looks like plain text except for the empty lines which VIM shows as `<200b>`

**Comment by shantamg** [comment]
And, for what it's worth, if I paste it into this very text field everything looks fine but I can tell that the zero width space is there because when I move the cursor with the arrow keys it takes an extra keystroke to get past that point.

**Comment by marijnh** [comment]
Right, both vim and github's textarea only take text, so they won't use the HTML content.

**Comment by adrianheine** [comment]
> Right, both vim and github's textarea only take text, so they won't use the HTML content.

And that text content shouldn't contain the zero-width space, should it?

**Comment by marijnh** [comment]
No, it shouldn't, which suggests that the native copy behavior is taking place. But on my Chrome, it isn't.

**Comment by adrianheine** [comment]
Right, I just wanted to point out that copying to vim is a valid, pretty minimal test. I can reproduce the issue on Linux with Chromium 58 with [your setup](https://www.ably.io/documentation/realtime/push/activate-subscribe).

**Comment by marijnh** [comment]
So what I'm doing is copy a bunch of text with blank lines from a CodeMirror instance, and if I go to vim and do 'i' 'ctrl-shift-v', it is pasted, but doesn't appear to contain zero-width spaces.

**Comment by phuongkt** [comment]
Let me take experence

**Comment by mattheworiordan** [comment]
> Update: Are these snippets rendered on the client or pre-rendered? @mattheworiordan

@adrianheine sorry for my lack of response, I had an email rule some time back that mistakenly deleted notifications :(

The snippets are rendered on the client.

---

It seems this issue is still affecting people, and not just people who use Xcode sadly. See https://twitter.com/PutMionyourteam/status/1023223425165352960.  When copying and pasting this code into vue.js templates, the [zero-width space](http://www.fileformat.info/info/unicode/char/200b/index.htm) is causing vue.js to bomb out.  

Does anyone know what the motivation for using a zero-width space is within CodeMirror?

**Comment by marijnh** [comment]
> Does anyone know what the motivation for using a zero-width space as opposed to say a non-breaking space is within CodeMirror?

It has zero width.

As I mentioned before, these are in the DOM, but there _should_ be a clear separation between what's in the DOM and what you get when you copy the text (where CodeMirror's copy handler gives you a nice, clean textual representation of the selected content).

Are you doing something like setting the editor to `readOnly: "nocursor"` or selecting _across_ it when copying? If so, yes, you will likely get an effect like this, but there's unfortunately no good way to override the copied text for some DOM content, you can only override the entire copy event, and CodeMirror only does that when it is focused.

**Comment by mattheworiordan** [comment]
Thanks for your response @marijnh 

> It has zero width.

Well sure, I realise that is what it is technically called.  However, why the need for a zero-width character when it does not render as zero-width when selected anyway.  See below example with the zero-width `&#8203;` character compared with a `&nbsp;` character.  Clearly the non breaking char is wider, but the zero-width one is definitely not zero-width.

![screen_capture_2018-07-31_10-23-53_am](https://user-images.githubusercontent.com/43789/43451037-edf4b48e-94ab-11e8-8a34-2fb1303b23aa.png)

> As I mentioned before, these are in the DOM, but there should be a clear separation between what's in the DOM and what you get when you copy the text (where CodeMirror's copy handler gives you a nice, clean textual representation of the selected content).

Ok, certainly sounds like that's the right approach.

> Are you doing something like setting the editor to readOnly: "nocursor"

Yes, would you mind explaining why that is an issue?  In the docs it states:

>> If the special value "nocursor" is given (instead of simply true), focusing of the editor is also disallowed.'

I did a `git blame` to work out when this was introduced, and it seems our devs used this setting from the start in 2015.  I'd love to understand why this setting is an issue if you wouldn't mind explaining.

I tried changing it on a local dev version of the rendered Code, and it made no difference.

> or selecting across it when copying?

Well no, I don't think people are copying across, but they are selecting part of the content.  They are copying & pasting the code they require, not the entire code block.  See example below with `readOnly: true`.

![screen_capture_2018-07-31_10-43-08_am](https://user-images.githubusercontent.com/43789/43452098-862e4dbc-94ae-11e8-992a-7a51ea421537.png)

>  If so, yes, you will likely get an effect like this, but there's unfortunately no good way to override the copied text for some DOM content

Well we could just `&nbsp;` in place of zero-width characters in the rendered DOM, and the problem would be solved.  This could just be a setting perhaps to enabled / disable this feature.

> you can only override the entire copy event, and CodeMirror only does that when it is focused.

Ok, not really sure what can be done then on our side :(

FYI. In the mean time, I am now running a selector to find these offending spans and manually replace them with non-breaking space spans.  Not ideal, but I believe this will work in the short term.  I'll get the affected user to confirm.

**Comment by marijnh** [comment]
> Yes, would you mind explaining why that is an issue?

Because, as I described, that won't allow the editor to be focused, and thus its clipboard event handlers won't run.

`readOnly=nocursor` has a bunch of issues like this, and will be dropped in the next API-breaking version.

> See example below with readOnly: true.

Your screenshots show that you're getting the DOM selection, not the editor's own selection, so I suspect something went wrong with your test and you're still using nocursor.

If you go to the online demos on codemirror.net, you'll see that the problem doesn't occur there.

> Not ideal, but I believe this will work in the short term.

Unlikely. This'll cause all kinds of problems—the editor does not support random DOM mutations in the content it rendered.

**Comment by mattheworiordan** [comment]
Thanks again @marijnh 

> If you go to the online demos on codemirror.net, you'll see that the problem doesn't occur there.

I have, and I don't see any read-only demos.  We use codemirror exclusively to render code as part of our page, not for editing, so our use case is perhaps not "standard".  

>> Not ideal, but I believe this will work in the short term.

> Unlikely. This'll cause all kinds of problems—the editor does not support random DOM mutations in the content it rendered.

Well it does work for us, because as above, we only use CodeMirror for rendering and styling code, not for editing, so the obvious complications related to modifying the DOM for editable code is not a problem.

**Comment by marijnh** [comment]
If you're just _rendering_ code, you might want to look into the [runmode addon](https://codemirror.net/doc/manual.html#addon_runmode), and generate simpler DOM that doesn't have any editing-related cruft in it.

**Comment by mattheworiordan** [comment]
Oooh, that looks interesting.  I'll raise an issue in our backlog and see if we can do a spike to see if we can swap that out.

Thanks!
